Disruptive upgrade

jsuchome opened the pull request:
**Why is this change necessary?**

Disruptive upgrade is needed for users without the setup that allows non-disruptive upgrade

**How does it address the issue?**

Implements the proposal described at https://etherpad.nue.suse.com/p/cloud-upgrade-disruptive


**For followup** 

- ~~Showing some nice status when doing paralel parts: https://github.com/crowbar/crowbar-core/pull/1133~~ **Merged**
- More paralelization (currently it is done only for OS upgrade of nodes)

jsuchome commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(8) S56P5 Implement disruptive upgrade (non-HA, one compute node, etc.)](https://trello.com/c/2B0jUWrs/334-8-s56p5-implement-disruptive-upgrade-non-ha-one-compute-node-etc)

Itxaka commented on chef/cookbooks/crowbar/templates/default/crowbar-evacuate-host.sh.erb:
```diff
@@ -47,7 +47,7 @@ fi
 # FIXME: here we need checks for potential errors during the evacuation
 
-while nova list --host "$host" --fields host 2>/dev/null | grep "$host"; do
+while nova list --all-tenants --status active --host "$host" --fields host 2>/dev/null | grep "$host"; do
```
Not sure about status active. What about instances that are shutting off? They are still running until they finally shut off so they should be taken into account?

jsuchome replied:
This is actually used during the live migration, where instances are not shutting down, but moving to different host. (So it's maybe confusing to have it in this PR).

Anyway, your comment probably still applies, do you have any idea how to make the check better? We need to make sure that the live-migration was successfully finished and the $host in question does not have any instances running (it could have some not-running though).

Itxaka replied:
Probably get all instances and filter with a whitelist for statuses that we know are troublesome? I cant remember the status of an instance migrating, but its probably not `active` so we could be going forward while there is still instances migrating :/

jsuchome replied:
@Itxaka check the new state, I'm grepping for MIGRATING status now

Itxaka commented on crowbar_framework/app/models/api/upgrade.rb:
```diff
@@ -535,6 +572,63 @@ def cancel
+      def upgrade_disruptive_controllers
+        save_upgrade_state("Entering disruptive upgrade of controller nodes")
+
+        # Go through all OpenStack barclamps (skip "pacemaker" and "openstack") ordered by run_order
```
Is this comment correct? Seems that it actually skips anything NOT on the openstack category?

jsuchome replied:
Here I'm not sure what was the meaning of "openstack" here, there's no barclamp like that... @rhafer ?

(It's true that we are skipping all nodes that do not have any openstack roles assigned here... that was not intentional, but I think it could be solved at the end of upgrade by a check for 'anything that is left')

rhafer replied:
> Here I'm not sure what was the meaning of "openstack" here, there's no barclamp like that... @rhafer ?

The ```prop.category != "OpenStack"``` takes information from the ```config/catalog.yml```. Which itself is generated from the various ```<barclampname>.yml``` in the git repositores. There you can define, that a certain barclamp is a "member" of another barclamp. ```prop.category != "OpenStack"``` matches all barclamps that are member of the "OpenStack" barclamp (yes, there is a barclamp called openstack, it doesn't do much and does really provide a UI)

> It's true that we are skipping all nodes that do not have any openstack roles assigned here... that was not intentional, but I think it could be solved at the end of upgrade by a check for 'anything that is left'

What would that be?

jsuchome replied:
So, you're saying that the check is correct or not?

Itxaka replied:
This is not correct or at least it does not sync with the comment
> Go through all OpenStack barclamps (skip "pacemaker" and "openstack") ordered by run_order

This is not skipping the openstack barclamp but its skipping any barclamps that are NOT in the openstack category, like `ntp` or `provisioner` which have the `crowbar` category. Which Im guessing its the end goal here, so the comment should change.

```
irb(main):021:0> bar = []
irb(main):022:0> BarclampCatalog.barclamps.each do |k, v|
irb(main):023:1* p = Proposal.where(barclamp: k).first
irb(main):024:1> next if p.nil?
irb(main):025:1> next if p.category != "OpenStack"
irb(main):026:1> next if k == "pacemaker"
irb(main):027:1> bar.push(k)
irb(main):028:1> end
irb(main):029:0> bar
=> ["nova", "keystone", "rabbitmq", "neutron", "cinder", "glance", "horizon", "database"]
```

jsuchome replied:
> This is not correct or at least it does not sync with the comment

That's what I tried to say earlier https://github.com/crowbar/crowbar-core/pull/1126#discussion_r104113114 . I believe the code is correct, and the comment should go, unless @rhafer (who wrote the original like) disagrees.

jsuchome replied:
@Itxaka oh, so once again, we are skipping barclamp named `openstack`, but that does not mean we are skipping other barclamps in `OpenStack` category

Itxaka commented on crowbar_framework/app/models/api/upgrade.rb:
```diff
@@ -535,6 +572,63 @@ def cancel
+            ).first
+            upgrade_cluster founder, cluster_env
+          else
+            # if cluster(s) and node(s) assigned to the role are already upgraded, skip to next role
```
Is this done here or its already coded in the `upgrade_one_node` function? `return if node.upgraded?`

jsuchome replied:
Yes, that's actually solved by `node.upgraded?` check later. (I should remove the comment)

jsuchome commented:
Needs rebase after merging https://github.com/crowbar/crowbar-core/pull/1129

jsuchome commented on crowbar_framework/app/models/api/upgrade.rb:
```diff
@@ -535,6 +572,81 @@ def cancel
+          proposal["deployment"][name]["element_order"].flatten.each do |el|
+            # skip nova compute nodes, they will be upgraded separately
+            next if el.start_with?("nova-compute") || elements[el].nil?
+            # skip cinder-volume nodes that also have a nova compute role assigned
```
We might want to skip also `ceilometer-agent` nodes...

jsuchome commented:
Rebased after the mode renaming

rhafer commented on crowbar_framework/app/models/api/upgrade.rb:
```diff
@@ -535,6 +572,81 @@ def cancel
+      end
+
+      # Go through active proposals and return elements with assigned roles in the right order
+      def upgradable_elements_of_proposals(proposals)
```
I somehow agree with Hound here that this is a bit to complex. I am currently experimenting a bit with this to see if we could make it a bit simpler.

rhafer replied:
In general this function seems to have an issue with mixed cluster/non-cluster setups. E.g. if keystone is deployed non-HA on a node that is a cluster member for some later barclamp (glance, neutron, cinder, ...) it won't upgrade the whole cluster.

I know this is not really a supported setup (and no sane person would install that it that way), but we might want to deal with it in  the disruptive upgrade path somehow. (Solution would be to load the node objects and check if they are a cluster member).

rhafer commented on crowbar_framework/app/models/api/upgrade.rb:
```diff
@@ -535,6 +537,83 @@ def cancel
+            next if el.start_with?("nova-compute") || elements[el].nil?
+            # skip some roles if they are assigned to compute nodes
+            if ["cinder-volume", "ceilometer-agent", "swift-storage"].include? el
+              elements[el].each do |node|
```
I think this won't work if e.g. ```cinder-volume``` is assigned to a cluster. In that case ```elements[el]``` will contain ```cluster:<name>```.  So it's probably better to expand the elements here. ```ServiceObject.expand_nodes_for_all``` should be able to do that.

jsuchome replied:
`cinder-volume` in cluster and compute node at the node that forms this cluster? That's probably broken... :-)

rhafer replied:
"unsupported" yes, But not sure if it's completely broken. And I don't think we currently prevent such a setup from being upgraded. (Maybe we should)

rhafer commented on crowbar_framework/app/models/api/upgrade.rb:
```diff
@@ -535,6 +537,83 @@ def cancel
+          elements = nova_prop["deployment"]["nova"]["elements"]
+          elements.each do |role, nodes|
+            next unless role.start_with?("nova-compute")
+            nodes.each { |node| compute_nodes[node] = 1 }
```
In compute-ha setups this is not a list of nodes but somethink like ```remotes:<name>```. Should also be expandable using ```ServiceObject.expand_nodes_for_all```.

But then again it seems compute-ha is using a completely different role ..., hm.

jsuchome replied:
Good point, but can we save it to compute-ha work? Probably some other code parts will be affected as well...

rhafer replied:
Sure

rhafer commented on crowbar_framework/app/models/api/upgrade.rb:
```diff
@@ -535,6 +537,83 @@ def cancel
+        nova_prop = proposals["nova"][:prop] || nil
+        unless nova_prop.nil?
+          elements = nova_prop["deployment"]["nova"]["elements"]
+          elements.each do |role, nodes|
```
Nit: Why not ```nova_prop["deployment"]["nova"]["elements"].each do |role, nodes|```?

jsuchome replied:
true

rhafer commented on crowbar_framework/app/models/api/upgrade.rb:
```diff
@@ -535,6 +537,83 @@ def cancel
+          next if ["pacemaker", "openstack"].include? name
+          prop = Proposal.where(barclamp: name).first
+          next if prop.nil? || !prop.active?
+          proposals[name] = { order: BarclampCatalog.run_order(name), prop: prop }
```
This still seems a bit weird to be. I think you don't need the order here. Just doing:

```proposals[name] = prop```

 here and sorting it later with:

```proposals = proposals.sort_by { |key, _unused| BarclampCatalog.run_order(key) }.to_h```

should do the trick. (Or sorting the barclamps already before loading the proposal and writing the ```proposal``` hash in a sorted fashion from the beginning as suggested in my initial comment.

rhafer commented on crowbar_framework/app/models/api/upgrade.rb:
```diff
@@ -535,6 +537,83 @@ def cancel
+        to_upgrade
+      end
+
+      def upgrade_disruptive_controllers
```
To allign with ```upgrade_nodes_disruptive``` I suggest to rename this to ```upgrade_controllers_disruptive```

rhafer commented on crowbar_framework/app/models/api/upgrade.rb:
```diff
@@ -535,6 +537,83 @@ def cancel
         upgrade_status.initialize_state
       end
 
+      def upgrade_nodes_disruptive(to_upgrade)
```
Please add a comment that this is really working with elements and not nodes. (Maybe even rename it).

Also I suggest to rename ```to_upgrade``` to ```elements_to_upgrade```

jsuchome replied:
will do

rhafer commented on crowbar_framework/app/models/api/upgrade.rb:
```diff
@@ -357,6 +357,32 @@ def target_platform(options = {})
+        return if out[:stdout].nil? || out[:stdout].empty?
+        hosts = out[:stdout].split.uniq.join("\n")
+        raise_services_error(
+          "There are some running instances at nodes\n#{hosts}."
```
```The following compute nodes still have instances running: ....```

rhafer commented on crowbar_framework/app/models/api/upgrade.rb:
```diff
@@ -378,7 +404,8 @@ def services
 
-        # FIXME: report an error if there is running instances and mode is disruptive
+        status = ::Crowbar::UpgradeStatus.new
+        check_for_running_instances if status.upgrade_mode == "disruptive"
```
Using just ```upgrade_mode``` should work here.

rhafer replied:
it's ```status.upgrade_mode == :normal``` (you also don't really need the ```status``` here as Api::Upgrade provides an ```upgrade_mode``` method. But I am repeating myself ;)

rhafer commented on crowbar_framework/spec/models/api/upgrade_spec.rb:
```diff
@@ -367,7 +379,175 @@
+
+      allow(Node).to(
+        receive(:find).with("roles:nova-compute-kvm").
+        and_return([Node.find_by_name("testing.crowbar.com")])
```
Would be good to really test this with multiple compute nodes. And verify that ```execute_scripts_and_wait_for_finish``` is actually called with multiple nodes passed.

rhafer commented on crowbar_framework/spec/models/api/upgrade_spec.rb:
```diff
@@ -367,7 +379,175 @@
+
+      allow_any_instance_of(Crowbar::UpgradeStatus).to receive(:progress).and_return(
+        suggested_upgrade_mode: :normal
+      )
```
Same as above: 
As ```Crowbar::UpgradeStatus.new.upgrade_mode``` is already tested elsewhere. IMO it's ok to change this to:

```
allow_any_instance_of(Crowbar::UpgradeStatus).to receive(:upgrade_status).and_return( :normal )
```

here and delete the next line.

rhafer commented on crowbar_framework/app/models/api/upgrade.rb:
```diff
@@ -1143,7 +1145,38 @@ def upgrade_all_compute_nodes(upgrade_mode = :non_disruptive)
       end
 
-      def upgrade_compute_nodes(virt)
+      def paralel_upgrade_compute_nodes(compute_nodes)
```
Typo: "paralel"

jsuchome commented on crowbar_framework/spec/models/api/upgrade_spec.rb:
```diff
@@ -367,7 +379,175 @@
+
+      allow_any_instance_of(Crowbar::UpgradeStatus).to receive(:progress).and_return(
+        suggested_upgrade_mode: :normal
+      )
```
you mean `allow_any_instance_of(Crowbar::UpgradeStatus).to receive(:upgrade_mode)` right?

rhafer replied:
Oops. Yeah.

jsuchome commented:
Rebased after the merge of https://github.com/crowbar/crowbar-core/pull/1133. Added a call of `save_nodes_state` into `parallel_upgrade_compute_nodes`

rhafer commented on crowbar_framework/app/models/api/upgrade.rb:
```diff
@@ -1143,7 +1145,39 @@ def upgrade_all_compute_nodes(upgrade_mode = :non_disruptive)
+            "Error while upgrading compute nodes. " + e.message
+          )
+        end
+        # FIXME: should we paralelize this as well?
```
Yeah, especially having the ```join_and_chef``` part  running in parallel could speed things up quite a bit. But it's ok to address that later IMO, just add it to the backlog.

jsuchome replied:
Also, let's keep in mind that for pure compute nodes, chef-client does not take that much time.

rhafer replied:
Well even if it just takes 30 seconds per node (and I am pretty sure it takes longer) it sums about to 25 minutes for 50 minutes. But as I said it's ok to address that later.

rhafer commented on crowbar_framework/app/models/api/upgrade.rb:
```diff
@@ -585,22 +585,31 @@ def cancel
-            # FIXME: could we do it in paralel?
-            upgrade_one_node element
+            node = ::Node.find_by_name(element)
+            if node["run_list_map"].key? "pacemaker-cluster-member"
```
IMO this code would better fit into ```upgradable_elements_of_proposals```. Where you would "just" replace the nodenane with the cluster element before inserting it into the result list.

And if you already load the nodes in ```upgradable_elements_of_proposals``` anyway you could also rework the check for computes a bit to also just look at the run_list_map.

jsuchome replied:
Hm, I'm not sure. I'm mostly working with lists in `upgradable_elements_of_proposals`, so loading a Node there seems like another level of complexity.

jsuchome commented:
Rebased after merging https://github.com/crowbar/crowbar-core/pull/1136

jsuchome commented:
Rebased after merge of https://github.com/crowbar/crowbar-core/pull/1134

jsuchome commented:
> My only concern is, once this gets in we need to remember to kill the manila-service vm for any 
> disruptive  CI jobs ...

Good point, but wasn't @abel-navarro already on it?

jsuchome commented on crowbar_framework/app/models/api/upgrade.rb:
```diff
@@ -370,6 +370,32 @@ def target_platform(options = {})
+      # Check if they are any instances running.
+      # It's necessary for disruptive upgrade, when user has to shut them down manually before
+      # proceeding with the services shutdown.
+      def check_for_running_instances
```
We'll need to add some check for this into our jobs - suspend running VM's before the services step. Including the manila one @Itxaka mentioned.
